src/aotcompile.cpp

```diff
@@ -1505,213 +1472,250 @@ void jl_dump_native_impl(void *native_code,
             CMModel,
             CodeGenOpt::Aggressive // -O3 TODO: respect command -O0 flag?
             ));
+    auto DL = jl_create_datalayout(*SourceTM);
+    std::string StackProtectorGuard;
+    unsigned OverrideStackAlignment;
+    data->M.withModuleDo([&](Module &M) {
+        StackProtectorGuard = M.getStackProtectorGuard().str();
+        OverrideStackAlignment = M.getOverrideStackAlignment();
+    });
 
+    auto compile = [&](Module &M, StringRef name, unsigned threads, auto module_released) {
+        return add_output(M, *SourceTM, name, threads, !!unopt_bc_fname, !!bc_fname, !!obj_fname, !!asm_fname, module_released);
+    };
 
-    std::vector<NewArchiveMember> bc_Archive;
-    std::vector<NewArchiveMember> obj_Archive;
-    std::vector<NewArchiveMember> asm_Archive;
-    std::vector<NewArchiveMember> unopt_bc_Archive;
-    std::vector<std::string> outputs;
-
-    // Reset the target triple to make sure it matches the new target machine
-    auto dataM = data->M.getModuleUnlocked();
-    dataM->setTargetTriple(SourceTM->getTargetTriple().str());
-    dataM->setDataLayout(jl_create_datalayout(*SourceTM));
-
-    Type *T_size = dataM->getDataLayout().getIntPtrType(Context);
-    Type *T_psize = T_size->getPointerTo();
+    SmallVector<AOTOutputs, 16> sysimg_outputs;
+    SmallVector<AOTOutputs, 16> data_outputs;
+    SmallVector<AOTOutputs, 16> metadata_outputs;
+    if (z) {
+        LLVMContext Context;
+        Module sysimgM("sysimg", Context);
+        sysimgM.setTargetTriple(TheTriple.str());
+        sysimgM.setDataLayout(DL);
+        sysimgM.setStackProtectorGuard(StackProtectorGuard);
+        sysimgM.setOverrideStackAlignment(OverrideStackAlignment);
+        Constant *data = ConstantDataArray::get(Context,
```

Comment:
I think you also now could move the call to `jl_create_system_image` here, and avoid needing all of that extra memory (for `z`) until after LLVM is completely done and after we've freed most of the llvm::Module objects.

Reply:
I suspect that change is more complicated, since we call `jl_create_native` as part of `jl_create_system_image`, and then I think we do a few more transformations on the module while generating `z`.

Reply:
Yeah, but in that case the change might make more sense too, since the current organization sounds complicated also